### ISSUE TITLE
Implement synchronized BiMaps.

### ISSUE THREAD

**Issue by motlin** [eclipse/eclipse-collections#25]
The API `MutableBiMap<K, V>.asSynchronized()` already exists, but [the implementation](https://github.com/eclipse/eclipse-collections/blob/7.0.0/eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java#L385) in `AbstractMutableBiMap` is unsupported.

**Comment by bhav0904** [comment]
`public class SynchronizedBiMap<K, V>
        extends AbstractMutableBiMap<K, V> implements MutableBiMap<K, V>, Externalizable`

Is this the correct signature for the synchronized BiMap?

**Comment by motlin** [comment]
I think that `SynchronizedBiMap` should not extend `AbstractMutableBiMap` because its methods are not thread-safe. It should either extend `AbstractSynchronizedMapIterable`, or a new abstract subclass of `AbstractSynchronizedMapIterable` but I'm not sure what that would be.

**Comment by bhav0904** [comment]
Something like:
`public abstract class AbstractSynchronizedMutableBiMap<K, V>
        extends AbstractSynchronizedMapIterable<K, V> implements MutableBiMap<K, V>`

where `AbstractSynchronizedMutableBiMap` would be similar to `AbstractMutableBiMap` but would have thread-safe implementations.  

But there is a problem with this design: `groupByUniqueKey` returns `MutableBiMap` in the `MutableBiMap` interface and returns `MutableMap` in `AbstractSynchronizedRichIterable`, which is causing a clash in my proposed `AbstractSynchronizedMutableBiMap`.

1. Why does groupByUniqueKey return `MutableBiMap` and not `MutableMap` in the `MutableBiMap` interface? GroupBy, by definition just needs to return a map (or a multimap) right? Where in it's specification does it say that it has to be a bidirectional map (even though the object invoking the method is a bidirectional map)?

2. If we decide to let MutableBiMap.groupByUniqueKey keep returning `MutableBiMap`, AbstarctSynchronizedMutableBiMap (or even SynchronizedMutableBIMap) cannot extend  `AbstractSynchronizedRichIterable`.

Thoughts?

**Comment by motlin** [comment]
Good questions. In `BiMap<K, V>`, both `K` and `V` are unique by definition. Here's the signature of `BiMap.groupByUniqueKey()`

```java
<VV> BiMap<VV, V> groupByUniqueKey(Function<? super V, ? extends VV> function);
```

The contract of `groupByUniqueKey` is that `VV` must also be unique. The result of `BiMap.groupByUniqueKey()` is a `MapIterable<VV, V>` where both `VV` and `V` are unique. We might as well say it's a `BiMap`. If we had `BiMaps` sorted on one or both types, they'd work here too.

It's a mistake that `AbstractSynchronizedRichIterable.groupByUniqueKey()` returns `MutableMap` instead of `MapIterable`. I noticed that too and fixed it in my fork in commit 2323819f52c1c43057aaf758f75b21b5b97f7161. It's mixed with other work that I'm not ready to merge in, and will eventually rebase, but you could cherry-pick that commit into your fork if you want to build on top of it.

**Comment by bhav0904** [comment]
Interesting. Thanks for your answers! 

I think I will build on top of your commit, and go with my `AbstractSynchronizedMutableBiMap<K, V>` plan. Does that sound good?

**Comment by motlin** [comment]
Yes, sounds good to me!

**Comment by bhav0904** [comment]
Craig, could you double-check your commit? It looks like you fixed the return type of `AbstractMutableMapIterable.groupByUniqueKey()` but not the return type of `AbstractSynchronizedRichIterable.groupByUniqueKey()`. I could fix `AbstractSynchronizedRichIterable.groupByUniqueKey()` to return MapIterable and commit that first.

**Comment by motlin** [comment]
They are in two separate commits. At the moment, the hash codes are ad28a7dc11df486f549cace57dc87846a7a18821 and e49cd9328ed775e3d75e85e327338c6131eab002.

**Comment by bhav0904** [comment]
Thanks Craig. I finally started working on Synchronized BiMaps. 

Firstly, I dropped the idea of an "abstract" Synchronized BiMap because it didn't really solve any purpose. My SynchronizedMutableBiMap looks like this now:

`public class SynchronizedMutableBiMap<K, V> extends AbstractSynchronizedMapIterable<K, V> implements MutableBiMap<K, V>, Serializable`

Secondly, I committed my changes but didn't submit a pull request because one unit test was failing and I wanted to discuss it with you. The failing test case is: `MutableMapIterableTest.keyPreservation()`. SynchronizedMutableBiMap.keysView() uses Lazy Iterable adapted keySet() under the hood. The keySet() in a HashBiMap uses the inverse's iterator, which means that if you call HashBiMap.keySet().iterator(), it will actually give you the iterator of the inverse, which makes sense.
But in this case, we expect the key to be preserved, which is actually the "value" of the inverse, causing it not to be preserved, and the test case fails. 
I wondered why it doesn't fail for HashBiMap, but realized that is is because HashBiMap doesn't use keySet() inside keysView(), which allows it not to delegate to the inverse. 
I feel that this is not a breach of contract in any way, it is just how the BiMap's keysView works. We could either: 
1. Drop the test for SynchronizedMutableBiMap, or
2. Modify the test just for SynchronizedMutableBiMap so that it doesn't use keysView() to get the keys

Let me know what you think.

Other points I wanted to raise:

1. `collect(Function<? super V, ? extends VV> function)` doesn't have a covariant override in MutableBiMap. I believe it should be:
`<VV> MutableBag<VV> collect(Function<? super V, ? extends VV> function)`. 
This is currently causing the collect method inside SynchronizedMutableBiMap to return RichIterable. I wasn't sure if I could include this fix in a minor version. 

2. Similarly for all collect<primitive> methods.

All of this is in my [commit](https://github.com/bhav0904/eclipse-collections/commit/9fea373231f5e190dd31816e9ef990ede66991ae). Please review when you get a chance and let me know what you think!

**Comment by motlin** [comment]
I'll take a look at your commit to understand the failing test. In the meantime, could you look at b007c996992f9d75e49e6c04e7c2aa8df63a0e08 and see if it addresses the covariant return type for `collect()`?

**Comment by motlin** [comment]
You've stumbled upon a real bug in HashBiMap. It's unrelated to keySet() and keysView(). I wrote the smallest tests I could to show the problem. Since it's a bug, maybe we can fix it on the 7.0.x branch. That's where I've committed the tests: 7a5cebf240243b6f46bc2e5738d4f1933be1d79b.

**Comment by motlin** [comment]
Here's the build showing the assertion failures.

https://travis-ci.org/motlin/eclipse-collections/jobs/107351816

**Comment by bhav0904** [comment]
Thanks Craig! I committed a fix to my 7.0.x branch and created a pull request (after cherry-picking your commit that added failing tests), but my [pull request](https://github.com/eclipse/eclipse-collections/pull/41) fails - is it because I'm trying to create a request on top of your commit?

**Comment by motlin** [comment]
No, it's due to committing without signing. You can click the red X for more details.

https://dev.eclipse.org/eclipse-webhook/services/status_details.php?id=56b901c55f56b

**Comment by bhav0904** [comment]
I signed off on my commit, but it looks like it is complaining that "The identity in the Signed-off-By must match the Author or the Committer." which is probably because of on the commits has you as the author.

**Comment by motlin** [comment]
I see what you mean. I think this means I need to do the cherry pick myself. Look for the changes in my fork of 7.0.x shortly.

```
$ git log --format=fuller 65eae42
commit 65eae42cb028622ddc875df33df6c0ca17672845 (upstream/pr/41/head, hindub/7.0.x)
Author:     Bhavana Hindupur <bhavana.hindupur@gmail.com>
AuthorDate: Mon Feb 8 14:52:58 2016 -0500
Commit:     Bhavana Hindupur <bhavana.hindupur@gmail.com>
CommitDate: Mon Feb 8 14:52:58 2016 -0500

    Fix memory leak in HashBiMap.
    
    Signed-off-by: Bhavana Hindupur <bhavana.hindupur@gmail.com>

commit b842a56aef429b4c629024f1c3fd03d514963815
Author:     Craig P. Motlin <craig.motlin@gs.com>
AuthorDate: Fri Feb 5 17:25:58 2016 -0500
Commit:     Bhavana Hindupur <bhavana.hindupur@gmail.com>
CommitDate: Mon Feb 8 13:04:48 2016 -0500

    Write tests showing that there is a memory leak in HashBiMap.
```

**Comment by motlin** [comment]
Oh, it's already available in my branch. Assuming you have a remote set up for me, you can rebase like this.

git fetch --all
git rebase --onto 7a5cebf240243b6f46bc2e5738d4f1933be1d79b b842a56aef429b4c629024f1c3fd03d514963815 7.0.x

**Comment by bhav0904** [comment]
I'm sorry, I had messed up my 7.0.x, so I recreated it and added a new [commit](https://github.com/bhav0904/eclipse-collections/commit/911ed1d96b9acb4de0c13be48af5745de536ba25) with the changes (along with the changes from your review comments). 

So do you now need to cherry pick this new commit and push it to the upstream 7.0.x?

**Comment by motlin** [comment]
No, I think the commands I suggested are still valid, assuming your local clone is fairly similar to your public fork.

**Comment by motlin** [comment]
I see you've already fixed your 7.0.x branch. I've reopened the pull request so that the Travis builds will kick off. Let's move any more conversation relevant to the bug in HashBiMap to #41.
